### ISSUE TITLE
Extra section for adobe source code pro font family.

### ISSUE THREAD

**Issue by hassanes** [bhilburn/powerlevel9k#591]
Since change from PR #577 it fixed Python icon while using `awesome-terminal-fonts` ,but it break python icon in `source-code-pro` user. 
![screenshot from 2017-08-06 18-54-49](https://user-images.githubusercontent.com/20126200/29006364-561fc796-7b18-11e7-9885-883117046e95.png)

I've looked at [How to install a font wiki](https://github.com/bhilburn/powerlevel9k/wiki/Install-Instructions#step-2-install-a-powerline-font). See option 2 (Use programmer font) and option 3 (Install Awesome-Powerline Fonts) share the same `POWERLEVEL9K_MODE='awesome-fontconfig'`. 

These two fonts glyph map are not compatible. See Python icon in (real Python logo) `awesome-terminal-fonts`  is mapped to `\ue63c` but Python icon (snake icon) `source-code-pro`  is mapped to `\U1F40D`.

Should we create extra `POWERLEVEL9K_MODE` for `source code pro` font family?

**Comment by bhilburn** [comment]
Hey @hassanes - Yeah, I think this is a good idea.

It's not actually clear to me what glyphs Source Code Pro actually includes, though. Source Code Pro doesn't, on it's own, include Powerline - you need the Powerline version of the font: https://github.com/powerline/fonts/tree/master/SourceCodePro

Further, even that version doesn't include the "Awesome Font" glyphs - you need yet a further patched version I think? One of these:

* https://github.com/ryanoasis/nerd-fonts/tree/master/patched-fonts/SourceCodePro/Regular/complete
* https://github.com/jldeen/bad-ass-terminal/blob/master/SourceCodePro%2BPowerline%2BAwesome%2BRegular.ttf

What version of the font are you using? I do want to make sure that we have an easy-to-use config for this, and if we need a new font section for it, then let's do it.

**Comment by hassanes** [comment]
I've installed `adobe-source-code-pro-font` and `fontawesome-fonts` via dnf, a Fedora package manager.
This is version of each fonts.
![screenshot from 2017-08-11 21-15-09](https://user-images.githubusercontent.com/20126200/29217021-adadcb28-7eda-11e7-8efb-3a8f8ef47e02.png)

**Comment by bhilburn** [comment]
Okay, let's get this rolling =)

I just pushed a new font configuration type, `source-code-pro`. Check out the `new-adobe-scp-iconset` branch to try it out.

I could use help setting the code points for the icons. There are a few different distributions of the Source Code Pro font family that are used:

1. Adobe Source Code Pro: https://github.com/adobe-fonts/source-code-pro
2. Source Code Pro + Powerline: https://github.com/powerline/fonts/tree/master/SourceCodePro
3. Nerd Fonts Sauce Code Pro (might be derived from #2?): https://github.com/ryanoasis/nerd-fonts/tree/master/patched-fonts/SourceCodePro/Regular
4. Awesome Fonts Sauce Code Powerline: https://github.com/gabrielelana/awesome-terminal-fonts/tree/patching-strategy/fonts
5. This random file that floats around. No idea where it came from, but it's referenced in many blog posts: https://github.com/bhilburn/dotfiles/blob/master/fonts/SourceCodePro%2BPowerline%2BAwesome%2BRegular.ttf

I tried the canonical Adobe Source Code Pro a while back, but it didn't have the powerline symbols at the time. Based on [this Issue in the Nerd Fonts](https://github.com/ryanoasis/nerd-fonts/issues/33), though, maybe that changed?

Regardless, the point is we need to figure out what font to use for code-point definitions. What would be best is to use the font that is packaged in OS default repos. 

@hassanes - I assume the Fedora repo is packaging the canonical Adobe Source Code Pro font? Can you figure out the source of the font, and does it include powerline symbols? If you have installed `awesome-terminal-fonts` via DNF, did it install the SCP version of those (#4 in the list above), too?

Pinging @gabrielelana and @ryanoasis in case they have input on how best to do this. Especially since many programs are starting to ship with defaults of SCP (it is [the default font for Spacemacs, for example](https://github.com/syl20bnr/spacemacs/blob/bd7ef98e4c35fd87538dd2a81356cc83f5fd02f3/doc/DOCUMENTATION.org#font)), it would be good not replicate font installations for the same font family.

**Comment by ryanoasis** [comment]
> Nerd Fonts Sauce Code Pro (might be derived from #2?)

Originally yes it was derived from the Powerline patched version (Sauce Code Pro), but you're correct that it is now derived from Source Code Pro (SCP) canonical.

---

SCP does include Powerline symbols since [Roman fonts version 2.010 and Italic fonts version 1.030 ~July 2015](https://github.com/adobe-fonts/source-code-pro/releases/tag/2.010R-ro%2F1.030R-it)

I checked out SCP canonical:

![selection of SCP](https://user-images.githubusercontent.com/8083459/29755214-9825debe-8b62-11e7-8456-90f81ec895a4.png)

Ignore the unicode codepoint hex. I did a compact encoding to view glyphs more easily.
The powerline glyphs _are_ included but there are _not_ a lot of additional glyphs that I can see.

Here is a view of all symbols:

![selection SCP all glyphs](https://user-images.githubusercontent.com/8083459/29755289-e98c3e64-8b63-11e7-868f-9d9c377f9e3d.png)

**Comment by bhilburn** [comment]
Thanks so much, @ryanoasis. That was extremely helpful.

After looking through the font-map, I agree - it really doesn't have much other than the basic symbols. It would work for our standard powerline font configuration (the default), but not for any Awesome / NerdFont settings. In short, I think #1 and #2 in my list above are now equivalent.

That leaves us with #3, from @ryanoasis, #4 from @gabrielelana, and #5, which has an unknown source (although I suspect it is the result of using some patching scripts from either/both Nerd Fonts and Awesome Fonts). The version in AwesomeFonts is 3 years old, so I'm assuming it is built on the Powerline patched version, where-as the NerdFonts has been updated for the more recent update to SCP.

@ryanoasis @gabrielelana - Do you have any insight into the relative compatibility between code-points in your two distributions of SCP? Are they roughly the same or entirely different?

**Comment by gabrielelana** [comment]
@bhilburn you are right #4 comes from a powerline patched older version of SCP. I guess that #5 is one of my experiments done around the time I ended up with #4, so you can consider #5 and #4 roughly the same

But I don't understand what's the question 😊 

> How best to do this?

If that's the question then with what strategy? Patching or merging?

If you want to patch then you need to ship another font and it would be better to patch the most recent version of SCP with most recent version of the glyphs you need (I can reuse my scripts if you want and see if the result is ok)

Otherwise you can let/instruct the users to merge whatever version of SCP they have installed with the icon fonts you need

As you know I prefer the second choice but you can't always go for it (ex. OSX)

**Comment by bhilburn** [comment]
@gabrielelana - Thanks for the response!

By my question, "How best to do this?", I meant how best to define the code-points in P9k? Thinking about it a bit further, though, the solution might be obvious:

Users who install #1 (or #2, which seem to be the same, now), can just use the default P9k config (powerline glyphs, no extra glyphs). But what about users who want the "Awesome Font" glyphs? They might install ASCP from your AwesomeFonts repo (#4) or from @ryanoasis NerdFonts repo (#3).

If ASCP in NerdFonts follows roughly the same code-points as the other NerdFonts, we can just tell them to use the `nerdfont-complete` configuration, which we already have. If they install the AwesomeFonts version, then perhaps the code-points will be the same as your other AwesomeFonts?

In short, maybe we don't need a special configuration for this at all, anymore. That's the way it's looking to me.

**Comment by gabrielelana** [comment]
@bhilburn I can confirm that all icon glyphs of all my patched fonts share the same codepoints

**Comment by bhilburn** [comment]
@gabrielelana - Fantastic. Thanks!

@ryanoasis - Thoughts on the above plan?

**Comment by ryanoasis** [comment]
@bhilburn Yeah it is good with me and it makes sense. Basically no special configuration needed?

Yes Nerd Fonts (the complete variations) will have the same consistent code points between the differing font variations.

The only changes between versions would be additions of new glyphs (non breaking changes).

Any moving of existing glyphs would be breaking changes and thus the major version would be bumped (unlikely but if it happened we'd do our best to notify other projects .. though the projects could just safely use a different version. This is unlikely as I see it now).

**Comment by bhilburn** [comment]
I am on a issue-cleaning spree. This issue hasn't had motion in quite some time, and so I am closing it out due to inactivity. If anyone would like to re-open it because they feel it is un-resolved and is something they care about, please feel free to re-raise the discussion!
